### PR TITLE
F144565: graphman client diff enhancements

### PR DESCRIPTION
**Pull request by graju256**
New features:-
- introducing new query generator
  - supporting all queries at the field level (ex: keyByAlias, trustedCertByName, etc)
  - summary = **all:summary**
- introducing new operation (**describe**) to describe queries
- introducing new operation (**slice**) to operate on single bundle
- now, you can override any part of configuration using params (--options.*, --gateways.*)
- support for multiple schemas (--options.**schema**)
- support for **enhanced** **diff** report/bundle generation.
- support for exporting/importing policies and services along with their policy revisions

### CODE REVIEW

File: modules/extn/graphman-extension-multiline-text-diff.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright ©  2024. Broadcom Inc. and/or its subsidiaries. All Rights Reserved.
+ */
+
+const diffExtn = require('diff');
+
+module.exports = {
+    /**
+     * Extension to compute multiline text differences.
```
Comment (uppoju) [inline modules/extn/graphman-extension-multiline-text-diff.js:9]: description looks to be different

File: modules/graphman-bundle.js [hunk 2]
```diff
@@ -1,48 +1,76 @@
...
+     * Callback will be invoked for every class of entities in the bundle. They will be invoked with the key, entities and type-info as arguments
+     * @param bundle input bundle
+     * @param knownEntitiesCallback callback function for the known entities
+     * @param unknownEntitiesCallback callback function for the unknown entities
+     */
+    forEach: function (bundle, knownEntitiesCallback, unknownEntitiesCallback) {
+        Object.entries(bundle).forEach(([key, entities]) => {
+            const typeInfo = graphman.typeInfoByPluralName(key);
+            if (typeInfo) {
+                if (entities.length) knownEntitiesCallback(key, entities, typeInfo);
```
Comment (uppoju) [inline modules/graphman-bundle.js:35]: should we have null check for knownEntitiesCallback as well ?

File: modules/graphman-bundle.js [hunk 3]
```diff
@@ -140,97 +182,205 @@ module.exports = {
...
+
+            if (!right.hasOwnProperty(key)) { // is field itself missing?
+                if (callback) {
+                    data.right = null;
+                    callback(data);
+                    equals = false;
+                } else {
+                    return false;
+                }
+            } else if (left[key] == null && right[key] != null) { // special case: are they both nulls?
```
Comment (uppoju) [inline modules/graphman-bundle.js:288]: condition differs from comment

Comment (graju256) [reply modules/graphman-bundle.js:288]: even though it doesn't appear this condition (serialization removes null fields), let's handle this properly.

Thanks.

File: modules/graphman-bundle.js (hunk 2 above)
Comment (graju256) [reply modules/graphman-bundle.js:35]: **knownEntitiesCallback** is mandatory.

File: modules/extn/graphman-extension-multiline-text-diff.js (hunk 1 above)
Comment (graju256) [reply modules/extn/graphman-extension-multiline-text-diff.js:9]: didn't understand!.

File: modules/graphman-bundle.js (hunk 3 above)
Comment (graju256) [reply modules/graphman-bundle.js]: @uppoju Addressed it.

File: modules/extn/graphman-extension-multiline-text-diff.js (hunk 1 above)
Comment (uppoju) [reply modules/extn/graphman-extension-multiline-text-diff.js:9]: functional arguments does not match with the javadoc

**Review by uppoju — APPROVED** [review]

File: modules/graphman-bundle.js (hunk 2 above)
Comment (uppoju) [reply modules/graphman-bundle.js:35]: OK

**Pull request merged**